python/descriptor.c

```diff
@@ -38,113 +42,919 @@
+    {"name", PyUpb_EnumValueDescriptor_GetName, NULL, "name"},
+    {"number", PyUpb_EnumValueDescriptor_GetNumber, NULL, "number"},
+    //{ "index", (getter)GetIndex, NULL, "index"},
+    {"type", PyUpb_EnumValueDescriptor_GetType, NULL, "index"},
+    {"has_options", PyUpb_EnumValueDescriptor_GetHasOptions, NULL,
+     "Has Options"},
+    {NULL}};
+
+static PyMethodDef PyUpb_EnumValueDescriptor_Methods[] = {
+    {
+        "GetOptions",
+        PyUpb_EnumValueDescriptor_GetOptions,
+        METH_NOARGS,
+    },
+    {NULL}};
+
+static PyType_Slot PyUpb_EnumValueDescriptor_Slots[] = {
+    DESCRIPTOR_BASE_SLOTS,
+    {Py_tp_methods, PyUpb_EnumValueDescriptor_Methods},
+    {Py_tp_getset, PyUpb_EnumValueDescriptor_Getters},
+    {0, NULL}};
+
+static PyType_Spec PyUpb_EnumValueDescriptor_Spec = {
+    PYUPB_MODULE_NAME ".EnumValueDescriptor",  // tp_name
+    sizeof(PyUpb_DescriptorBase),              // tp_basicsize
+    0,                                         // tp_itemsize
+    Py_TPFLAGS_DEFAULT,                        // tp_flags
+    PyUpb_EnumValueDescriptor_Slots,
+};
 
 // -----------------------------------------------------------------------------
 // FieldDescriptor
 // -----------------------------------------------------------------------------
 
-static PyObject *PyUpb_FieldDescriptor_GetType(PyUpb_DescriptorBase *self,
-                                               void *closure) {
+const upb_fielddef* PyUpb_FieldDescriptor_GetDef(PyObject* _self) {
+  PyUpb_DescriptorBase* self =
+      PyUpb_DescriptorBase_Check(_self, kPyUpb_FieldDescriptor);
+  return self ? self->def : NULL;
+}
+
+PyObject* PyUpb_FieldDescriptor_Get(const upb_fielddef* field) {
+  const upb_filedef* file = upb_fielddef_file(field);
+  return PyUpb_DescriptorBase_Get(kPyUpb_FieldDescriptor, field, file);
+}
+
+static PyObject* PyUpb_FieldDescriptor_GetFullName(PyUpb_DescriptorBase* self,
+                                                   void* closure) {
+  return PyUnicode_FromString(upb_fielddef_fullname(self->def));
+}
+
+static PyObject* PyUpb_FieldDescriptor_GetName(PyUpb_DescriptorBase* self,
+                                               void* closure) {
+  return PyUnicode_FromString(upb_fielddef_name(self->def));
+}
+
+static PyObject* PyUpb_FieldDescriptor_GetCamelcaseName(
+    PyUpb_DescriptorBase* self, void* closure) {
+  // TODO: Ok to use jsonname here?
+  return PyUnicode_FromString(upb_fielddef_jsonname(self->def));
+}
+
+static PyObject* PyUpb_FieldDescriptor_GetJsonName(PyUpb_DescriptorBase* self,
+                                                   void* closure) {
+  return PyUnicode_FromString(upb_fielddef_jsonname(self->def));
+}
+
+static PyObject* PyUpb_FieldDescriptor_GetFile(PyUpb_DescriptorBase* self,
+                                               void* closure) {
+  const upb_filedef* file = upb_fielddef_file(self->def);
+  if (file) {
+    return PyUpb_FileDescriptor_Get(file);
+  } else {
+    Py_RETURN_NONE;
```

Comment:
I generally prefer the structure of

```
if (!file) Py_RETURN_NONE;

return ...
```

as that follows the common idiom of bailing from error early and having the mainline continue, so the code requires less churn if it evolves in the future

Reply:
Done.